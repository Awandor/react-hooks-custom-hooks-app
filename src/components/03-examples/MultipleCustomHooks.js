import { useCounter } from "../../hooks/useCounter";
import { useFetch } from "../../hooks/useFetch";

export const MultipleCustomHooks = () => {

    const { counter, increment } = useCounter( 1 );

    console.log( 'Counter', counter );

    const url = `https://www.breakingbadapi.com/api/quotes/${counter}`;

    const { loading, data, error } = useFetch( url );

    // console.log( loading );

    // console.log( 'data:', data[0].quote ); // error

    // El problema es que el estado inicial de data es null y null[0] da error, hay que validar

    const { author, quote } = !!data && data[0]; // !!data -> si es true entonces retorna data[0]

    // Transformamos null en booleano null = null, !null = true, !!null = false
    // Si es falso no continúa evaluando y author y quote tendrán valor undefined

    return (
        <div className="container">
            <h1>Breaking Bad Quotes</h1>
            <hr />

            {
                loading ? ( <div className="alert alert-info text-center">Loading...</div> )
                    : ( <blockquote className="blockquote text-right"><p className="mb-0">{quote}</p><footer className="blockquote-footer">{author}</footer></blockquote> )
            }

            <button type="button" className="btn btn-outline-primary" onClick={() => {

                increment( 1 );

            }}>Next quote</button>

        </div>
    );

};
