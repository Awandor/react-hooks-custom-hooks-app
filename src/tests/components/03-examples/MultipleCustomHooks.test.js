import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks'
import { shallow } from 'enzyme';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

jest.mock( '../../../hooks/useFetch' );
jest.mock( '../../../hooks/useCounter' );

// Podríamos falsear los datos de otros custom Hooks de la misma manera


describe( 'Prueba de <MultipleCustomHooks />', () => {

    // Como useCounter no lo vamos a alterar, definimos el mock antes de cada test con beforeEach

    beforeEach( () => {

        useCounter.mockReturnValue( {
            counter: 10,
            increment: () => { }
        } );
    
    } );

    test( 'debe mostrarse correctamente', () => {

        useFetch.mockReturnValue( {
            data: null,
            loading: true,
            error: null
        } );

        const wrapper = shallow( <MultipleCustomHooks /> );

        expect( wrapper ).toMatchSnapshot();

    } );

    test( 'debe mostrar los datos', () => {

        // Vamos a empezar por el custom Hook useFetch, pero vamos a decirle a Jest que no queremos comprobar
        // si useFetch funciona, ya lo hemos comprobado, sólo nos interesa la información que va a retornar

        // Para ello vamos a pasar useFetch por un mock, esto hace que en ESTE ARCHIVO cuando use useFetch
        // va a utilizar unaimplementación que me voy a inventar, como es en este archivo tengo que implementarlo
        // también en la prueba anterior

        useFetch.mockReturnValue( {
            data: [ {
                author: 'Dan Häggblom',
                quote: 'Hola mundo'
            } ],
            loading: false,
            error: null
        } );

        const wrapper = shallow( <MultipleCustomHooks /> );

        // En vez de tomar una foto podemos buscar por elementos HTML

        expect( wrapper.find( '.alert' ).exists() ).toBeFalsy();
        expect( wrapper.find( '.mb-0' ).exists() ).toBeTruthy();
        expect( wrapper.find( '.mb-0' ).text().trim() ).toBe( 'Hola mundo' );
        expect( wrapper.find( 'footer' ).text().trim() ).toBe( 'Dan Häggblom' );

        // expect( wrapper ).toMatchSnapshot();

    } );

} );
