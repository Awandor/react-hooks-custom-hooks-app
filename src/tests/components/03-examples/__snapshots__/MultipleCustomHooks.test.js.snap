// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba de <MultipleCustomHooks /> debe mostrarse correctamente 1`] = `
<div
  className="container"
>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-outline-primary"
    onClick={[Function]}
    type="button"
  >
    Next quote
  </button>
</div>
`;
